src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,242 @@
+                            }
+
+                            var calledMethods = new ConcurrentDictionary<IMethodSymbol, bool>();
+                            callGraph.TryAdd(methodSymbol, calledMethods);
+
+                            operationBlockStartAnalysisContext.RegisterOperationAction(operationContext =>
+                            {
+                                callGraph[methodSymbol].TryAdd((operationContext.Operation as IInvocationOperation).TargetMethod, true);
+                            }, OperationKind.Invocation);
+                        });
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        (CompilationAnalysisContext compilationAnalysisContext) =>
+                        {
+                            var visited = new HashSet<IMethodSymbol>();
+                            var results = new Dictionary<IMethodSymbol, HashSet<IMethodSymbol>>();
+
+                            foreach (var kvp in callGraph)
+                            {
+                                var methodSymbol = kvp.Key;
+                                // Determine if the method is called automatically when an object is deserialized.
+                                // This includes methods with OnDeserializing attribute, method with OnDeserialized attribute, deserialization callbacks as well as cleanup/dispose calls.
+                                var parameters = methodSymbol.GetParameters();
+                                var flagHasDeserializeAttributes = attributeTypeSymbols.Length != 0
+                                    && attributeTypeSymbols.Any(s => methodSymbol.HasAttribute(s))
+                                    && parameters.Length == 1
+                                    && parameters[0].Type.Equals(streamingContextTypeSymbol);
+                                var flagImplementOnDeserializationMethod = methodSymbol.IsOnDeserializationImplementation(IDeserializationCallbackTypeSymbol);
+                                var flagImplementDisposeMethod = methodSymbol.IsDisposeImplementation(compilation);
+                                var flagIsFinalizer = methodSymbol.IsFinalizer();
+
+                                if (!flagHasDeserializeAttributes && !flagImplementOnDeserializationMethod && !flagImplementDisposeMethod && !flagIsFinalizer)
+                                {
+                                    continue;
+                                }
+
+                                FindCalledDangerousMethod(methodSymbol, visited, results);
+
+                                foreach (var result in results[methodSymbol])
+                                {
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        methodSymbol.CreateDiagnostic(
+                                            Rule,
+                                            methodSymbol.ContainingType.MetadataName,
```

Comment:
Any specific reason of using `MetadataName` instead of `Name` here? I believe both should be the same given method is declared in source.